### ISSUE TITLE
cppcoreguidelines-pro-type-member-init false positive with delegate constructor in template

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#37250]
|  |  |
| --- | --- |
| Bugzilla Link | [37902](https://llvm.org/bz37902) |
| Version | unspecified |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @carlosgalvezp,@EugeneZelenko,@fabianwolff,@fberger,@gnzlbg,@tomlankhorst |

## Extended Description 
cppcoreguidelines-pro-type-member-init incorrectly triggers here:

template <typename T>
struct A {
  A() : A(42) {}
  explicit A(int value) : value_(value) {}

  int value_;
};

$ clang-tidy -extra-arg=-std=c++11 -checks='*' foo.cc 
2 warnings generated.
foo.cc:3:3: warning: constructor does not initialize these fields: value_ [cppcoreguidelines-pro-type-member-init]
  A() : A(42) {}
  ^
$

I see this problem with both clang-tidy 5.0.1 and 6.0.0, both with and without -extra-arg=-std=c++11.

I was only able to reproduce this problem in templates.  In this minimized example, note that the template parameter type T is not even used!  When "template <typename T>" is removed, which does not change the behavior of the struct, the problem goes away.  (This isn't related to *unused* template parameters -- the problem also occurs when T is used normally.)

See also llvm/llvm-project#31579  , which describes a false positive for a non-template delegated constructor.  I was unable to reproduce the problem with the example given there.  (But, adding "template <typename T>" to that example caused the problem.)

**Comment by carlosgalvezp** [comment]
Hi! Are there any updates on this issue?

**Comment by gnbond** [comment]
This is also an issue with template constructors to a non-template class, even when the delegation is to a non-template constructor:
```
#include <cstddef>

class S {
    std::size_t m_size;

 public: 
    S(std::size_t size) : m_size{ size } {}

    template<typename C>
    S(const C& c) : S( c.size()) {}
};

#include <vector>

const std::vector<int> vi{};

const S s{vi};

```
```
clang-tidy t.cpp 
16630 warnings generated.
t.cpp:10:5: warning: constructor does not initialize these fields: m_size [cppcoreguidelines-pro-type-member-init]
    S(const C& c) : S( c.size()) {}
    ^
Suppressed 16629 warnings (16629 in non-user code).

```
(clang-tidy 11.1 which is the latest I have access to on this old Mac)

**Comment by PiotrZSL** [comment]
Workaround on clang-tidy level: https://reviews.llvm.org/D157367
